### PR TITLE
Surface Meshing Simplification

### PR DESCRIPTION
**Pull request by Jmeyer1292**
This pull request is under active development. Please do not merge. I just want it out there for review.

This pull request changes the method of (mesh) triangulation used for surfaces. Once planes of interest are identified, points belonging to them are reprojected onto the plane, the concave hull of all of them is taken, and that is meshed using the ear clipping algorithm. 

This appears to screw up something with normal calculations stored inside the mesh point cloud, so I (dangerously) disabled the check. For the test cases, I've used it still seems to function fine and the plane normals are re-calculated anyway.

Thanks for your patience.

### CODE REVIEW

**Comment by VictorLamoine** [comment]
The C++11 issue looks like https://github.com/PointCloudLibrary/pcl/issues/619.
This happens only with the pre-compiled packages, not if you compile PCL by hand. (see https://github.com/ros-industrial-consortium/godel/issues/38)

File: godel_plugins/CMakeLists.txt [hunk 1]
```diff
@@ -11,7 +11,7 @@ find_package(catkin REQUIRED COMPONENTS
 )
 
 # c++11
-ADD_DEFINITIONS("-std=c++0x")
+#ADD_DEFINITIONS("-std=c++0x")
```
Comment (jrgnicho) [inline godel_plugins/CMakeLists.txt]: Does this still need to be eliminated? Wasn't this an issue in debug build?

File: godel_process_path_generation/src/mesh_importer.cpp [hunk 2]
```diff
@@ -327,11 +327,11 @@ bool MeshImporter::computePlaneCoefficients(Cloud::ConstPtr cloud, Eigen::Vector
...
-  if (actual_normal.dot(expected_normal) < std::cos(seg.getEpsAngle()))
-  {
-    ROS_WARN("RANSAC plane normal out of tolerance! cosines: %f / %f", actual_normal.dot(expected_normal), std::cos(seg.getEpsAngle()));
-    return false;
-  }
+  //if (actual_normal.dot(expected_normal) < std::cos(seg.getEpsAngle()))
+  //{
+  //  ROS_WARN("RANSAC plane normal out of tolerance! cosines: %f / %f", actual_normal.dot(expected_normal), std::cos(seg.getEpsAngle()));
+  //  return false;
+  //}
```
Comment (jrgnicho) [inline godel_process_path_generation/src/mesh_importer.cpp]: Commented out code should be removed

File: godel_surface_detection/CMakeLists.txt [hunk 3]
```diff
@@ -19,7 +19,7 @@ find_package(catkin REQUIRED COMPONENTS
 find_package(PCL 1.7 REQUIRED COMPONENTS common io)
 find_package(Boost REQUIRED COMPONENTS system)
 
-ADD_DEFINITIONS("-std=c++0x")
+#ADD_DEFINITIONS("-std=c++0x")
```
Comment (jrgnicho) [inline godel_surface_detection/CMakeLists.txt:22]: Same as above

File: godel_surface_detection/src/detection/surface_detection.cpp [hunk 4]
```diff
@@ -764,39 +772,6 @@ bool SurfaceDetection::apply_kdtree_radius_search(const Cloud& query_points,cons
...
-
-	// triangulation
-	pcl::GreedyProjectionTriangulation<pcl::PointNormal> gt;
-	gt.setConsistentVertexOrdering(true);
-	gt.setSearchRadius(params_.tr_search_radius);
-	gt.setMu(params_.tr_mu);
-	gt.setMaximumNearestNeighbors(params_.tr_max_nearest_neighbors);
-	gt.setMaximumSurfaceAngle(params_.tr_max_surface_angle);
-	gt.setMinimumAngle(params_.tr_min_angle);
-	gt.setMaximumAngle(params_.tr_max_angle);
```
Comment (jrgnicho) [inline godel_surface_detection/src/detection/surface_detection.cpp:789]: There are some members that are only used by this fast triangulation routine ("tr_" prefix).  Since they are no longer being used they should be removed.  However this isn't a trivial task because some gui components in godel_plugins assume that they exists.  Thus this should probably be addressed in a separate PR.

File: godel_surface_detection/src/detection/surface_detection.cpp [hunk 5]
```diff
@@ -870,5 +845,59 @@ bool SurfaceDetection::apply_tabletop_segmentation(const Cloud& cloud_in,Cloud&
...
+{
+	pcl::ModelCoefficients::Ptr coeff_ptr(new pcl::ModelCoefficients());
+	pcl::PointIndices::Ptr plane_inliers_ptr(new pcl::PointIndices());
+	pcl::SACSegmentation<pcl::PointXYZ> seg;
+
+	// setting segmentation options
+	seg.setOptimizeCoefficients(true);
+	seg.setModelType(pcl::SACMODEL_PLANE);
+	seg.setMethodType(pcl::SAC_RANSAC);
+	seg.setMaxIterations(1000);
```
Comment (jrgnicho) [inline godel_surface_detection/src/detection/surface_detection.cpp]: Hardcoded constant should be placed in a static global variable.

File: godel_surface_detection/src/detection/surface_detection.cpp [hunk 6]
```diff
@@ -870,5 +845,59 @@ bool SurfaceDetection::apply_tabletop_segmentation(const Cloud& cloud_in,Cloud&
...
+	proj.setModelType (pcl::SACMODEL_PLANE);
+	proj.setInputCloud (out.makeShared());
+	proj.setModelCoefficients(coeff_ptr);
+	proj.filter (out);
+}
+
+bool SurfaceDetection::apply_concave_hull(const Cloud& in, pcl::PolygonMesh& mesh)
+{
+	pcl::PolygonMesh::Ptr mesh_ptr (new pcl::PolygonMesh);
+	static int count = 0;
```
Comment (jrgnicho) [inline godel_surface_detection/src/detection/surface_detection.cpp]: static unused variable

File: godel_surface_detection/src/detection/surface_detection.cpp [hunk 7]
```diff
@@ -870,5 +845,59 @@ bool SurfaceDetection::apply_tabletop_segmentation(const Cloud& cloud_in,Cloud&
...
+	pcl::ConcaveHull<pcl::PointXYZ> chull;
+	chull.setInputCloud (in.makeShared());
+	chull.setAlpha (0.1);
```
Comment (jrgnicho) [inline godel_surface_detection/src/detection/surface_detection.cpp]: hardcoded constant

**Comment by jrgnicho** [comment]
Just minor things.  +1 for merge after addressing these comments.  An issue should be created for the removal of the fast triangulation related code.

File: godel_plugins/CMakeLists.txt (hunk 1 above)
Comment (Jmeyer1292) [inline godel_plugins/CMakeLists.txt]: While I don't think it needs to be eliminated, it's not necessarily needed either.  Would you prefer I delete the flag all-together or switch it back on?

Comment (jrgnicho) [inline godel_plugins/CMakeLists.txt]: I don't see why keep it there then.  Let's remove it for now.

File: godel_surface_detection/src/detection/surface_detection.cpp (hunk 4 above)
Comment (Jmeyer1292) [inline godel_surface_detection/src/detection/surface_detection.cpp:789]: Got it.

**Comment by jrgnicho** [comment]
+1 for merge
